### PR TITLE
Custom prefix, optional versioning based on multiple commit messages

### LINKED ISSUE #72: Add Prefix to Tags
Hey,
First and foremost, great work with this library. I ran into an issue where I wanted to add a prefix to the tag besides "v". Is there any way we can make that change?
Thanks

### LINKED ISSUE #226: Double v bug when combining WITH_V and INITIAL_VERSION
If `WITH_V` is `TRUE` and `INITIAL_VERSION` includes a `v` prefix (eg `v0.0.0`), these values clash when calculating the initial tag version.

Solution options:
- Update the README with instructions to not put a prefix in `INITIAL_VERSION`. This is more susceptible to user error but is probably sufficient.
- Correctly handle prefixes in `INITIAL_VERSION`. This is more challenging to implement, particularly for custom prefixes.

A log excerpt showing the error is shown below:
```
2022-12-20T07:29:52.3716389Z *** CONFIGURATION ***
2022-12-20T07:29:52.3717187Z 	DEFAULT_BUMP: patch
2022-12-20T07:29:52.3717794Z 	WITH_V: true
2022-12-20T07:29:52.3719490Z 	RELEASE_BRANCHES: master,main
2022-12-20T07:29:52.3719962Z 	CUSTOM_TAG: 
2022-12-20T07:29:52.3720338Z 	SOURCE: .
2022-12-20T07:29:52.3720687Z 	DRY_RUN: false
2022-12-20T07:29:52.3818744Z 	INITIAL_VERSION: v0.0.0
2022-12-20T07:29:52.3822578Z 	TAG_CONTEXT: repo
2022-12-20T07:29:52.3823355Z 	PRERELEASE: false
2022-12-20T07:29:52.3824268Z 	PRERELEASE_SUFFIX: beta
2022-12-20T07:29:52.3824581Z 	VERBOSE: false
2022-12-20T07:29:52.3824849Z 	MAJOR_STRING_TOKEN: /major
2022-12-20T07:29:52.3825347Z 	MINOR_STRING_TOKEN: /minor
2022-12-20T07:29:52.3825601Z 	PATCH_STRING_TOKEN: /patch
2022-12-20T07:29:52.3826568Z 	NONE_STRING_TOKEN: /none
2022-12-20T07:29:52.3827238Z pre_release = false
2022-12-20T07:29:52.8054346Z From https://github.com/danrs/test-semver
2022-12-20T07:29:52.8054776Z  * [new branch]      bump       -> origin/bump
2022-12-20T07:29:52.8154378Z fatal: ambiguous argument 'vv0.0.0': unknown revision or path not in the working tree.
2022-12-20T07:29:52.8154925Z Use '--' to separate paths from revisions, like this:
2022-12-20T07:29:52.8155325Z 'git <command> [<revision>...] -- [<file>...]'
2022-12-20T07:29:52.8199340Z Last commit message: change default bump to patch
2022-12-20T07:29:52.9220706Z Bumping tag vv0.0.0 - New tag v
2022-12-20T07:29:52.9727144Z 2022-12-20T07:29:52Z: **pushing tag v to repo danrs/test-semver
```

### PR DESCRIPTION
**Pull request by L-Sypniewski**
**Added a possibility to use a custom prefix**

Added a possibility to define an optional PREFIX parameter to be added before x.y.z version number.
Removed WITH_V parameter - omitting PREFIX results in the same behaviour as setting WITH_V to false in a previous versions.

**Multiple commit messages are taken into account when bumping a version number**

New version number is based on all commit messages between a defined commit (HEAD by default) and a last tag with a given prefix. 
This behavior is disabled by default, it can be toggled on with `USE_LAST_COMMIT_ONLY` parameter set to  `false`.

**Added support for pull requests**
Specifying latest branch commit is useful when using this action for pull requests - one can set environment variable as follows: `BRANCH_LATEST_COMMIT: ${{ github.event.pull_request.head.sha }}` to calculate a new version basing on commits from a given PR.

The x.y.z precedence is preserved, so if there are three commits with `#minor` tags and two with `#major` then only two `#major` tags are taken into consideration - the new version is `(x+2).0.0` in that case.

Resolves #72 
Resolves #226 
Closes #95

### CODE REVIEW

**Review by plchampigny — CHANGES_REQUESTED** [review]

File: entrypoint.sh [hunk 1]
```diff
@@ -129,21 +192,30 @@ if $pre_release
...
 
 # set outputs
+new_tag_without_prefix=${new#"$prefix"}
+echo "New tag without prefix: $new_tag_without_prefix"
+echo "New tag: $new"
+echo "Prefix: $prefix"
+echo -e "Part incremented: $part\n\n"
+
 echo ::set-output name=new_tag::$new
+echo ::set-output name=new_tag_without_prefix::$new_tag_without_prefix
```
[the remark this answers is not in the document]
Comment (bjartek) [reply entrypoint.sh:206]: I need this aswell :)

File: entrypoint.sh [hunk 2]
```diff
@@ -4,7 +4,7 @@ set -o pipefail
 
 # config
 default_semvar_bump=${DEFAULT_BUMP:-minor}
-with_v=${WITH_V:-false}
+prefix=${PREFIX}
```
[the remark this answers is not in the document]
Comment (L-Sypniewski) [reply entrypoint.sh:7]: Good point, I'll implement this according to your suggestion

File: action.yml [hunk 3]
```diff
@@ -7,8 +7,10 @@ runs:
 outputs:
   new_tag:
     description: 'Generated tag'
+  new_tag_without_prefix:
+    description: 'Generated tag without specified prefix'
   tag:
-    description: 'The latest tag after running this action'
+    description: 'The latest tag before running this action'
```
[the remark this answers is not in the document]
Comment (L-Sypniewski) [reply action.yml:13]: It's a behavior change

File: entrypoint.sh [hunk 4]
```diff
@@ -47,12 +50,12 @@ git fetch --tags
...
+        tag=$(git for-each-ref --sort=-v:refname --format '%(refname)' | cut -d / -f 3- | grep -E "^($prefix)?[0-9]+.[0-9]+.[0-9]+$" | head -n1)
+        pre_tag=$(git for-each-ref --sort=-v:refname --format '%(refname)' | cut -d / -f 3- | grep -E "^($prefix)?[0-9]+.[0-9]+.[0-9]+(-$suffix.[0-9]+)?$" | head -n1)
         ;;
     *branch*) 
-        tag=$(git tag --list --merged HEAD --sort=-v:refname | grep -E "^v?[0-9]+.[0-9]+.[0-9]+$" | head -n1)
-        pre_tag=$(git tag --list --merged HEAD --sort=-v:refname | grep -E "^v?[0-9]+.[0-9]+.[0-9]+(-$suffix.[0-9]+)?$" | head -n1)
+        tag=$(git tag --list --merged HEAD --sort=-v:refname | grep -E "^($prefix)?[0-9]+.[0-9]+.[0-9]+$" | head -n1)
+        pre_tag=$(git tag --list --merged HEAD --sort=-v:refname | grep -E "^($prefix)?[0-9]+.[0-9]+.[0-9]+(-$suffix.[0-9]+)?$" | head -n1)
         ;;
     * ) echo "Unrecognised context"; exit 1;;
```
[the remark this answers is not in the document]
Comment (L-Sypniewski) [reply entrypoint.sh:60]: Thanks for pointing this out, I'll investigate this.

File: entrypoint.sh [hunk 5]
```diff
@@ -79,24 +82,84 @@ if [ "$tag_commit" == "$commit" ]; then
     exit 0
 fi
 
-# echo log if verbose is wanted
+# calculate new tag
```
[the remark this answers is not in the document]
Comment (L-Sypniewski) [reply entrypoint.sh:85]: Good point, I'll add such a switch. Any suggestion for naming it concisely?

[the remark this answers is not in the document]
Comment (L-Sypniewski) [reply entrypoint.sh:117]: Fixed, added `USE_LAST_COMMIT_ONLY` variable

**Comment by raphaelreyna** [comment]
Are there any plans on moving forward with this?
Adding custom prefixes would allow for this action to be used with Go submodules which would be pretty sweet :+1:

File: README.md [hunk 6]
```diff
@@ -38,23 +39,27 @@ _NOTE: set the fetch-depth for `actions/checkout@v2` to be sure you retrieve all
...
-- **CUSTOM_TAG** _(optional)_ - Set a custom tag, useful when generating tag based on f.ex FROM image in a docker image. **Setting this tag will invalidate any other settings set!**
-- **SOURCE** _(optional)_ - Operate on a relative path under $GITHUB_WORKSPACE.
-- **DRY_RUN** _(optional)_ - Determine the next version without tagging the branch. The workflow can use the outputs `new_tag` and `tag` in subsequent steps. Possible values are `true` and `false` (default).
-- **INITIAL_VERSION** _(optional)_ - Set initial version before bump. Default `0.0.0`.
-- **TAG_CONTEXT** _(optional)_ - Set the context of the previous tag. Possible values are `repo` (default) or `branch`.
-- **PRERELEASE_SUFFIX** _(optional)_ - Suffix for your prerelease versions, `beta` by default. Note this will only be used if a prerelease branch.
-- **VERBOSE** _(optional)_ - Print git logs. For some projects these logs may be very large. Possible values are `true` (default) and `false`.
+* **GITHUB_TOKEN** ***(required)*** - Required for permission to tag the repo.
```
Comment (wheelerlaw) [inline README.md:42]: What is the point of changing the `-` to `*`? They both render the same way.

**Comment by msummers-nr** [comment]
Will the custom prefix PR be accepted? Or some flavor of it?

I have a very pressing need to use a tag prefix other than `v`

Thanks

File: entrypoint.sh [hunk 7]
```diff
@@ -13,12 +16,15 @@ initial_version=${INITIAL_VERSION:-0.0.0}
 tag_context=${TAG_CONTEXT:-repo}
 suffix=${PRERELEASE_SUFFIX:-beta}
 verbose=${VERBOSE:-true}
+branch_latest_commit=$BRANCH_LATEST_COMMIT
+use_last_commit_only=${USE_LAST_COMMIT_ONLY:-true}
 
-cd ${GITHUB_WORKSPACE}/${source}
+cd "${GITHUB_WORKSPACE}"/"${source}"
```
Comment (sbe-arg) [inline entrypoint.sh:22]: "a"/"b" ? Why not "a/b" shellcheck

**Review by sbe-arg — COMMENTED** [review]
Don't disagree with this but id probably warn some breaking changes in the readme for existing used not having the with_v var set and using default.

File: entrypoint.sh (hunk 7 above)
Comment (L-Sypniewski) [reply entrypoint.sh:22]: @sbe-arg What breaking changes you refer to? As for not having `with_v` var set the behaviour is the same as it had been before - the param is optional with the default value of 'false'

Comment (L-Sypniewski) [reply entrypoint.sh]: Fixed

**Review by justin-newman — APPROVED** [review]
.

**Comment by justin-newman** [comment]
@anothrNick Could we get this PR merged in sometime soon?

**Comment by L-Sypniewski** [comment]
@sbe-arg  I'll have a look at this today in the evening (European time)

**Comment by tautschnig** [comment]
@sammcj Would you mind elaborating as to why this one was closed?

**Comment by sammcj** [comment]
@tautschnig due to unresolved merged commits since April.

**Comment by L-Sypniewski** [comment]
@sbe-arg I would need a couple of days to add more test cases and resolve the conflicts as I'd like to cover as many scenarios as I can in the tests. Besides that, I have discovered a several bugs after adding more tests so I need some time to fix them.

**Comment by sammcj** [comment]
No worries at all - totally my fault for closing it early 😅

**Comment by sbe-arg** [comment]
Thanks for addressing the conflicts. Ill review in the next few days.

**Comment by L-Sypniewski** [comment]
@sbe-arg May you hold on for now? As I'm adding the tests I keep finding small bugs. How about me pinging you once everything's resolved?

**Comment by sbe-arg** [comment]
@L-Sypniewski no rush from me I like the concept of this pr, take your time, ping me or @sammcj when you think is ready.
And many thanks for adding more tests and making it reliable!

**Comment by danrs** [comment]
@L-Sypniewski If you've not encountered this already, there is an unexpected interaction if a prefix is included in the `INITIAL_VERSION` string. I've documented it in #226

**Comment by L-Sypniewski** [comment]
@danrs  Thanks for letting me know, I'll add a tests for this and fix the issue if needed.
Here's a link to branch I'm working on if anyone wants to have a look what tests and fixes I've added so far: https://github.com/L-Sypniewski/github-tag-action/tree/add-tests

Changes from the linked branch would be merged to the branch for this PR as soon as I finish all the tests and bugfixes - I'm working on a separate branch to not pollute this PR.

**Comment by L-Sypniewski** [comment]
@sbe-arg I have added the tests and resolved the conflicts, the PR is ready to be reviewed.
I'd appreciate suggestions especially regarding organizing the test related code - it think those tests are valuable, but I they are hard to read.

@danrs #226 issue has been addressed and there's a testcase for it in this PR

**Comment by sbe-arg** [comment]
Thanks for sorting the conflicts.

We broke the flow for so many people in the last few prs specially related to history of commits. Adding the new last commit flags conflicts with the changes merged in the 1.61 1.60 and 1.59. Can you please remove the commits history and adjust it with the existing changes?

I do support the prefix idea I think is needed but I think it requires testing and is probably a great excuse for v2.

And thanks a million for the test scripts.

Cc @sammcj when you have time to have a look as well.

**Comment by sbe-arg** [comment]
We introduced the `BRANCH_HISTORY`

**Comment by sammcj** [comment]
What's happening with this one folks?

**Comment by sbe-arg** [comment]
I'm tempted to close this one and the author can reopen as necesary.

**Pull request closed without merging**